libraries/chain/account_evaluator.cpp

```diff
@@ -352,19 +356,51 @@ void_result account_update_evaluator::do_apply( const account_update_operation&
 
    bool sa_after = acnt->has_special_authority();
 
+#ifndef ASSET_BALANCE_SORTED
+   flat_set<asset_id_type> assets_after = acnt->get_top_n_control_assets();
+
+   const auto& special_meta = d.get_special_assets_meta_object();
+#endif
+
    if( sa_before && (!sa_after) )
    {
       const auto& sa_idx = d.get_index_type< special_authority_index >().indices().get<by_account>();
       auto sa_it = sa_idx.find( o.account );
       assert( sa_it != sa_idx.end() );
       d.remove( *sa_it );
+#ifndef ASSET_BALANCE_SORTED
+      // From next maintenance interval, need to stop updating this account's active/owner.
+      // Here we check if each asset is added earlier in this maintenance interval,
+      //   if yes, remove it from "added"; otherwise, add it to "removed".
+      d.modify( special_meta, [&assets_before](special_assets_meta_object& p) {
```

Comment:
Here we are modifying a persisted data object. Could this cause problems if you run witness_node with ASSET_BALANCE_SORTED and later without (or vice-versa, or yes-no-yes, or no-yes-no)? I have not tested these situations, just typing out loud.

If it does cause problems, it seems the use of this feature will be rare. So are the symptoms easily detectable? If so, we could [log a message | force a replay]. We should also document the consequences.

Reply:
Good questions.

1. when to replay

Currently only `DB_VERSION` is used to detect whether need a replay.

Ideally the node should automatically detect a) whether db schema changed E.G. caused by code changes or compile-time option changes and b) whether certain options in `config.ini` changed. But probably need too much work to make it perfect.

For this issue, a quick solution is to define `DB_VERSION` to different values when using different compile-time option.

2. automatically replay or not

The main concern is that replay takes long time. In case when a node operator accidentally ran a node with an incompatible data directory, being forced to replay is painful. So it makes some sense if the node refuses to start and shows an error message when incompatible db schema is detected.

IMHO these questions worth a new issue, or perhaps two issues.

@pmconrad @oxarbitrage your opinion?

Reply:
> For this issue, a quick solution is to define `DB_VERSION` to different values when using different compile-time options.

Done.

Reply:
IMO the quick solution is good enough.